Userland/Applications/PixelPaint/Tools/TextTool.cpp

```diff
@@ -0,0 +1,295 @@
+
+        // ImageEditor scale is a float, but we are working with int and IntRects.
+        auto scaled_cursor_x = static_cast<int>(ceilf(m_editor->scale() * static_cast<float>(editor_cursor_rect.x())));
+        auto scaled_cursor_y = static_cast<int>(ceilf(m_editor->scale() * static_cast<float>(editor_cursor_rect.y())));
+        auto scaled_cursor_width = static_cast<int>(ceilf(m_editor->scale() * static_cast<float>(editor_cursor_rect.width())));
+        auto scaled_cursor_height = static_cast<int>(ceilf(m_editor->scale() * static_cast<float>(editor_cursor_rect.height())));
+
+        auto scaled_cursor_rect = Gfx::IntRect { scaled_cursor_x + text_location.x(),
+            scaled_cursor_y + text_location.y(),
+            scaled_cursor_width,
+            scaled_cursor_height };
+        painter.fill_rect(scaled_cursor_rect, m_text_color);
+    }
+}
+
+void TextTool::apply_text_to_layer()
+{
+    auto layer = m_editor->active_layer();
+    GUI::Painter painter(layer->get_scratch_edited_bitmap());
+
+    auto demo_text = m_text_editor->text();
+    auto text_width = m_selected_font->width(demo_text);
+    auto text_height = m_selected_font->preferred_line_height() * static_cast<int>(m_text_editor->line_count());
+
+    painter.set_font(*m_selected_font);
+    auto text_rect = Gfx::Rect<int>(m_add_text_position, { text_width, text_height });
+    painter.draw_text(text_rect, demo_text, Gfx::TextAlignment::TopLeft, m_text_color);
+    m_editor->did_complete_action(tool_name());
+    layer->did_modify_bitmap(text_rect);
+}
+void TextTool::reset_tool()
+{
+    // This puts the tool back into initial state between text additions (except for selected font/color)
+    m_text_input_is_active = false;
+    m_is_dragging = false;
+    m_mouse_is_over_text = false;
+    m_text_editor->set_text(""sv);
+    m_editor->image().selection().end_interactive_selection();
+    m_cursor_blink_timer->stop();
+    m_editor->update();
+    m_editor->update_tool_cursor();
+}
+bool TextTool::on_keydown(GUI::KeyEvent& event)
+{
+    if (!m_text_input_is_active)
+        return false;
+
+    // Cancels current text entry
+    if (event.key() == Key_Escape) {
+        reset_tool();
+        return true;
+    }
+
+    // A plain Return is treated as accepting the current state and rasterizing to the layer.
+    // For multi-line text Shift + Enter will add new lines.
+    if (event.modifiers() == Mod_None && event.key() == Key_Return) {
+        apply_text_to_layer();
+        reset_tool();
+        return true;
+    }
+
+    // Pass the key event off to our TextEditor subclass which handles all text entry features like
+    // caret navigation, backspace/delete, etc.
+    m_text_editor->handle_keyevent({}, event);
+    m_editor->update();
+    return true;
+}
+
+Variant<Gfx::StandardCursor, NonnullRefPtr<Gfx::Bitmap>> TextTool::cursor()
+{
+    if (m_mouse_is_over_text)
+        return Gfx::StandardCursor::Move;
+    else
```

Comment:
Pls drop this else.

Reply:
Not sure I understand why to drop the else. In the event you are not over the text we want the cursor to be the arrow.

Reply:
Doh! I get it now. No need for the else, just have it be 

```
    if (m_mouse_is_over_text)
        return Gfx::StandardCursor::Move;
    
    return Gfx::StandardCursor::Arrow;
    ```